fixes customizer logic in the GsonGrpcJsonMarshallerBuilder

AnyRoad opened the pull request:
Motivation:

 - JsonFormat.Parser and JsonFormat.Printer classes are immutable and call of any method in customizer of the GsonGrpcJsonMarshallerBuilder creates new object.
 - Because of that result of the customer is ignored so we need a way to use new object created in the customizer.

Modifications:

- Replaces Consumer<T> which does not return anything with the Function<T, T> customizer which returns customized object.

Result:
 - We can actually customize JsonFormat.Parser and JsonFormat.Printer using the customizer object.

<!--
Visit this URL to learn more about how to write a pull request description:
https://armeria.dev/community/developer-guide#how-to-write-pull-request-description
-->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/line/armeria?pullRequest=6146) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/line/armeria?pullRequest=6146) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/line/armeria?pullRequest=6146) it.</sub>

minwoox commented:
Hi, @AnyRoad! Thanks for the PR!
Would you mind:
- Signing CLA
- Adding test cases that verities the differences?
- Adding a new method instead of changing the signature of the existing method which breaks the compatibility?

AnyRoad commented:
> Hi, @AnyRoad! Thanks for the PR! Would you mind:
> 
>     * Signing CLA
> 
>     * Adding test cases that verities the differences?
> 
>     * Adding a new method instead of changing the signature of the existing method which breaks the compatibility?

Hi @minwoox !
I've signed the CLA, added unit tests and reverted back the original methods. Since they do not work properly I've marked them as deprecated.

trustin commented on grpc/src/main/java/com/linecorp/armeria/common/grpc/GsonGrpcJsonMarshallerBuilder.java:
```diff
@@ -32,40 +33,98 @@
+                return parser;
+            });
         }
         return this;
```
How about logging a warning message when this method is called so that people can avoid this pitfall at all? e.g.

```suggestion
        if (!loggedJsonParserCustomizerWarning) {
            logger.warn("{}.jsonParserCustomizer(Consumer) is buggy; use jsonParserCustomizer(Function).",
                        getClass().getSimpleName());
            loggedJsonParserCustomizerWarning = true;
        }
        return this;
```

(`loggedJsonParserCustomizerWarning` could be `private static boolean`.)

AnyRoad replied:
good idea, added logging https://github.com/line/armeria/pull/6146/commits/0812f5a2c828b8f5878b40d2c5a3b6e93a86ddee

trustin commented on grpc/src/main/java/com/linecorp/armeria/common/grpc/GsonGrpcJsonMarshallerBuilder.java:
```diff
@@ -32,40 +33,98 @@
+                return printer;
+            });
+        }
+        return this;
```
Ditto - we could do something similar to the above comment here for `jsonPrinterCustomizer(Consumer)`.

AnyRoad replied:
added logging here as well https://github.com/line/armeria/pull/6146/commits/0812f5a2c828b8f5878b40d2c5a3b6e93a86ddee

trustin reviewed (commented):
Thanks a lot, @AnyRoad! Left a couple minor comments about logging.

trustin commented on grpc/src/main/java/com/linecorp/armeria/common/grpc/GsonGrpcJsonMarshallerBuilder.java:
```diff
@@ -30,42 +34,117 @@
     public GsonGrpcJsonMarshallerBuilder jsonPrinterCustomizer(
             Consumer<? super JsonFormat.Printer> jsonPrinterCustomizer) {
+        if (!loggedJsonPrinterCustomizerWarning) {
+            logger.warn("{}.jsonPrinterCustomizer(Consumer) does not work as expected," +
+                            "use jsonPrinterCustomizer(Function).",
+                    getClass().getSimpleName());
```
Let me fix the indentation:
```suggestion
            logger.warn("{}.jsonPrinterCustomizer(Consumer) does not work as expected; " +
                        "use jsonPrinterCustomizer(Function).",
                        getClass().getSimpleName());
```

AnyRoad replied:
Thanks! Intellij haven't formated it and there was no linter error so I assumed that indentation is ok 😃

trustin replied:
No worries. 😄

trustin commented on grpc/src/main/java/com/linecorp/armeria/common/grpc/GsonGrpcJsonMarshallerBuilder.java:
```diff
@@ -30,42 +34,117 @@
     public GsonGrpcJsonMarshallerBuilder jsonParserCustomizer(
             Consumer<? super JsonFormat.Parser> jsonParserCustomizer) {
+        if (!loggedJsonParserCustomizerWarning) {
+            logger.warn("{}.jsonParserCustomizer(Consumer) does not work as expected, " +
+                            "use jsonParserCustomizer(Function).",
+                    getClass().getSimpleName());
```
Let me fix the indentation:
```suggestion
            logger.warn("{}.jsonParserCustomizer(Consumer) does not work as expected, " +
                        "use jsonParserCustomizer(Function).",
                        getClass().getSimpleName());
```

trustin reviewed (approved):
Nice work! Thank you! 🙇

minwoox reviewed (approved):
Thanks! @AnyRoad 👍 👍 👍
